core/templates/dev/head/pages/exploration-editor-page/feedback-tab/services/thread-data.service.ts

```diff
@@ -45,153 +43,141 @@ angular.module('oppia').factory('ThreadDataService', [
-              _data.suggestionThreads.push(suggestionThread);
-              break;
-            }
-          }
-        }
-      }).then(onSuccess);
+    var _setFeedbackThreadFromBackendDict = function(backendDict) {
+      var thread = FeedbackThreadObjectFactory.createFromBackendDict(
+        backendDict);
+      if (!_threadsById.hasOwnProperty(thread.threadId)) {
+        _threadsById[thread.threadId] = thread;
+      } else {
+        thread = angular.copy(thread, _threadsById[thread.threadId]);
+      }
+      return thread;
     };
 
-    var _fetchMessages = function(threadId) {
-      return $http.get(_THREAD_HANDLER_PREFIX + threadId).then(function(res) {
-        getThreadById(threadId).setMessages(res.data.messages);
-      });
+    var _setSuggestionThreadFromBackendDict = function(backendDict) {
+      var thread = SuggestionThreadObjectFactory.createFromBackendDict(
+        backendDict);
+      if (!_threadsById.hasOwnProperty(thread.threadId)) {
+        _threadsById[thread.threadId] = thread;
+      } else {
+        thread = angular.copy(thread, _threadsById[thread.threadId]);
+      }
+      return thread;
     };
 
     return {
       data: _data,
       getData: function() {
         return _data;
       },
-      fetchThreads: function(onSuccess) {
-        return _fetchThreads(onSuccess);
+      getThread: function(threadId) {
+        return _threadsById[threadId];
+      },
+      fetchThreads: function() {
+        return $http.get(_THREAD_LIST_HANDLER_URL).then(response => {
+          _data.feedbackThreads = response.data.feedback_thread_dicts.map(
+            _setFeedbackThreadFromBackendDict);
+          _data.suggestionThreads = response.data.suggestion_thread_dicts.map(
+            _setSuggestionThreadFromBackendDict);
+          return _data;
+        });
       },
       fetchMessages: function(threadId) {
-        return _fetchMessages(threadId);
+        var thread = this.getThread(threadId);
+        if (!thread) {
+          return $q.reject('Can not fetch messages of nonexistent thread.');
+        }
+
+        return $http.get(_THREAD_HANDLER_PREFIX + threadId).then(response => {
+          thread.setMessages(response.data.messages);
+        });
       },
       fetchFeedbackStats: function() {
-        return $http.get(_FEEDBACK_STATS_HANDLER_URL).then(function(response) {
+        return $http.get(_FEEDBACK_STATS_HANDLER_URL).then(response => {
           _openThreadsCount = response.data.num_open_threads;
         });
       },
       getOpenThreadsCount: function() {
         return _openThreadsCount;
       },
-      createNewThread: function(newSubject, newText, onSuccess) {
+      createNewThread: function(newSubject, newText) {
         return $http.post(_THREAD_LIST_HANDLER_URL, {
           state_name: null,
           subject: newSubject,
-          text: newText
-        }).then(function() {
+          text: newText,
+        }).then(
+          this.fetchThreads
+        ).then(() => {
```

Comment:
One question -- will the nested "then" commands result in issues if the user interacts with the page while the async call is in process?

Also, what is the rationale for introducing an extra then(), rather than updating the frontend and populating lazily? Ditto in other places.